src/rgw/store/sfs/multipart.cc

```diff
@@ -134,7 +118,143 @@ int SFSMultipartUpload::complete(
+
+  auto parts_it = parts.cbegin();
+  auto etags_it = part_etags.cbegin();
+
+  outobj->metadata_change_version_state(store, ObjectState::WRITING);
+  for (; parts_it != parts.cend() && etags_it != part_etags.cend();
+        ++parts_it, ++etags_it) {
+    if (parts_it->first != etags_it->first) {
+      // mismatch part num
+      lsfs_dout(dpp, 0) << "mismatch part num, expected: " << parts_it->first
+                        << ", got " << etags_it->first << dendl;
+      return -ERR_INVALID_PART;
+    }
+
+    auto part = parts_it->second;
+    auto part_it_etag = rgw_string_unquote(etags_it->second);
+    if (part->etag != part_it_etag) {
+      lsfs_dout(dpp, 0) << "mismatch part etag, expected: " << part->etag
+                        << ", got " << part_it_etag << dendl;
+      return -ERR_INVALID_PART;
+    }
+
+    char part_etag[CEPH_CRYPTO_MD5_DIGESTSIZE];
+    hex_to_buf(part->etag.c_str(), part_etag, CEPH_CRYPTO_MD5_DIGESTSIZE);
+    hash.Update((const unsigned char *) part_etag, sizeof(part_etag));
+
+    // copy multipart object to object at offset
+    std::filesystem::path partpath =
+      store->get_data_path() / part->objref->path.to_path();
+
+    ceph_assert(std::filesystem::exists(partpath));
+    ceph_assert(std::filesystem::file_size(partpath) == part->len);
+
+    lsfs_dout(dpp, 10) << "read part " << parts_it->first << " from "
+                       << partpath << ", size: " << part->len << dendl;
+    std::ifstream part_in{partpath, std::ios::binary};
+
+    uint64_t read_len = 0;
+    uint64_t block_size = 4194304;  // 4MB
+    char *buf = (char *) std::malloc(block_size);
+    while (read_len < part->len) {
+      std::memset(buf, 0, block_size);
+      uint64_t to_read = std::min((part->len - read_len), block_size);
+      part_in.seekg(read_len);
+      if (!part_in.read(buf, to_read)) {
+        lsfs_dout(dpp, 0) << "error reading part " << parts_it->first
+                          << " from " << partpath
+                          << ", bytes: " << to_read << dendl;
+        return -ERR_INVALID_PART;
+      }
+
+      auto cur_offset = out.tellp();
+      out.seekp(0);
+      if (!out.write(buf, to_read)) {
+        lsfs_dout(dpp, 0) << "error writing part " << parts_it->first
+                          << " to " << outpath
+                          << ", bytes: " << to_read
+                          << ", offset: " << cur_offset << dendl;
+        return -ERR_INVALID_PART;
+      }
+      read_len += to_read;
+      ofs += to_read;
+      accounted_size += to_read;
+    }
+    std::free(buf);
+    lsfs_dout(dpp, 10) << "copied part " << parts_it->first
+                       << ", accounted: " << accounted_size
+                       << ", offset: " << ofs
+                       << dendl;
+    part_in.close();
+  }
+
+  out.flush();
+  out.close();
+
+  char final_etag[CEPH_CRYPTO_MD5_DIGESTSIZE];
+  char final_etag_str[(CEPH_CRYPTO_MD5_DIGESTSIZE * 2) + 16];
```

Comment:
what is 16?
I'm not a big fan of magic numbers :)

Reply:
this code was mostly lifted from dbstore's implementation, and at the time it felt it made sense. The 16 would seem to be meant to hold the `-%lld` below.
The more I think about it, the less sense it makes though. Yes, it is meant to hold `-%lld`, but that should either be 21 bytes (`-` plus 19 from the `LLONG_MAX` plus `\0`), or it should be 7 (`-` plus the number of characters in `10000` -- the maximum number of multipart parts --, plus `\0`).

I'll ensure we use the latter limits, and assert if something is weirdly wrong.